src/gtk/overlay.cpp

```diff
@@ -0,0 +1,222 @@
+
+    m_rect = wxRect(x, y, width, height);
+
+    int clientWidth, clientHeight;
+    m_window->GetClientSize(&clientWidth, &clientHeight);
+
+    if ( m_surface )
+    {
+        gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+
+        return;
+    }
+
+    // Create the surface:
+
+    GtkWidget* const overlay = GetOverlayWidget(m_window);
+    wxASSERT_MSG( overlay, "Overlay is not enabled for m_window" );
+
+    m_surface = gtk_drawing_area_new();
+    gtk_widget_set_has_window(m_surface, false);
+    gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+    gtk_widget_set_halign (m_surface, GTK_ALIGN_START);
+    gtk_widget_set_valign (m_surface, GTK_ALIGN_START);
+    gtk_widget_set_vexpand (m_surface, TRUE);
+    gtk_widget_set_hexpand (m_surface, TRUE);
+    gtk_widget_set_can_focus(m_surface, false);
+
+    gtk_overlay_add_overlay( GTK_OVERLAY(overlay), m_surface );
+    gtk_overlay_set_overlay_pass_through( GTK_OVERLAY(overlay), m_surface, true );
+
+    g_signal_connect(m_surface, "draw", G_CALLBACK(wxgtk_draw_overlay), this);
+
+    g_object_ref(m_surface);
+    gtk_widget_show(m_surface);
+}
+
+void wxOverlayImpl::BeginDrawing(wxDC *dc)
+{
+    wxCHECK_RET( dc, "NULL dc pointer" );
+
+    wxClientDCImpl* const
+        dcimpl = static_cast<wxClientDCImpl*>(dc->GetImpl());
+
+    wxCHECK_RET( dcimpl, wxS("DC is not a wxClientDC!") );
+
+    if ( m_cairoSurface )
+        cairo_surface_destroy(m_cairoSurface);
+
+    m_cairoContext = static_cast<cairo_t*>(dcimpl->GetCairoContext());
+    cairo_push_group(m_cairoContext);
+
+    m_cairoSurface = cairo_get_group_target(m_cairoContext);
+    cairo_surface_reference(m_cairoSurface);
+}
+
+void wxOverlayImpl::EndDrawing(wxDC *WXUNUSED(dc))
+{
+    if ( m_cairoContext )
+    {
+        cairo_pop_group_to_source(m_cairoContext);
+        cairo_surface_set_device_offset(m_cairoSurface, 0, 0);
+
+        m_window->RefreshRect(m_rect);
+    }
+}
+
+void wxOverlayImpl::Clear(wxDC *WXUNUSED(dc))
+{
+    if ( m_cairoContext )
+    {
+#if wxDEBUG_GTK_OVERLAY
+        cairo_set_source_rgb(m_cairoContext, 0, 0, 1); // blue
+        cairo_paint_with_alpha(m_cairoContext, 0.5);
+#else
+        GtkStyleContext* styleContext = gtk_widget_get_style_context(m_surface);
+        gtk_render_background(styleContext, m_cairoContext,
+                              0, 0, m_rect.width, m_rect.height);
+#endif
+    }
+}
+
+void wxOverlayImpl::Reset()
+{
+    if ( m_window )
+    {
+        // Leave the old region represented by m_rect in a consistent state by
+        // invalidating it here. this is necessary when (for exemple) the caret
+        // moved to a new position.
+        m_window->RefreshRect(m_rect);
+        m_window->Unbind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+        m_window = NULL;
+    }
+}
+
+void wxOverlayImpl::OnSize(wxSizeEvent& event)
+{
+    Reset();
+    event.Skip(); // wxCaret sample crashes without this!
```

Comment:
Comment is a bit superfluous, you're supposed to skip size events unless you explicitly want to prevent the default handling from taking place, which is not the case here.